### ISSUE TITLE
Last token logits truncation causes logit-target misalignment

### ISSUE THREAD

**Issue by giladturok** [apple/ml-diffucoder#15]
**Bug:** The  `get_logits()` function in `coupled_grpo.py` incorrectly truncates the last token's logits, causing a systematic off-by-one error in logit-target alignment during training.

**Root Cause:** In [`get_logits()`](https://github.com/apple/ml-diffucoder/blob/f735e01376b752b70356efe45192f31499bf30dc/src/open_r1/coupled_grpo.py#L281-L287), the function returns truncated logits:
```python
return logits[:, :-1]  # Shape: [num_iterations * 3 * batch_size, seq_len, vocab_size]
```
This truncation shifts all subsequent logit-target alignments in [_get_per_token_logp()](https://github.com/apple/ml-diffucoder/blob/f735e01376b752b70356efe45192f31499bf30dc/src/open_r1/coupled_grpo.py#L290):
```python
completion_logits = logits[:, -logits_to_keep:, :]    # Misaligned due to truncation
completion_targets = expanded_input[:, -logits_to_keep:]  # Original target positions
```

**Impact:** This causes each logit to be trained against the wrong target token. Not only is this incorrect, but it produces NaNs gradients on my machine and may furthermore affect the ability of reward functions to parse answers correctly.

**Proposed Fix:** Remove the truncation: `return logits` instead of `return logits[:, :-1]`.

**Question:** Is this truncation intentional for a reason I'm missing?

This truncation is common in auto-regressive LLMs -- where logits of token `i` predict the next token `i +1` -- but I don't think this occurs in diffusion LLMs -- where logits of noisy token `i` at timestep `t` predicts clean token `i` at timestep `0`.

**Comment by summmeer** [comment]
Hi @giladturok Thank you for pointing this out! Yes, this truncation is common in auto-regressive LLMs, and usually not for diffusion LLMs. But in our DiffuCoder case, our model is adapted from AR LLM - so technically we will keep this shift operation. The details of this adaptation approach are in our previous paper: Scaling Diffusion Language Models via Adaptation from Autoregressive Models https://github.com/HKUNLP/DiffuLLaMA 

If you're using dLLMs trained from scratch (e.g., LLaDA), remove this shift; if you're using dLLMs like DiffuCoder/Dream/DreamCoder developed by adaptation, I'm afraid that this shift operarion is necessary.

**Comment by giladturok** [comment]
Thanks for the clarification -- makes sense! I was indeed using LLaDA and was encountering many gradient NaNs because of this logit truncation. 

Somehow, I missed that DiffuCoder was built upon an auto-regressive model.
